[Frontend] Add chunked processing to handle long inputs in embedding models

x22x22 opened the pull request:
Original pr
https://github.com/vllm-project/vllm/pull/20837

The previous submission was not merged for too long, resulting in too many conflicts, so I'm resubmitting it.

@noooop
@hmellor
@maxdebayser

x22x22 committed:
The latest update introduces new text embedding examples and service scripts, incorporating chunk processing capabilities to handle exceptionally long inputs. The README documentation has been revised to provide comprehensive instructions on usage methods and configuration options.

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
修复合并多模态处理器参数的逻辑，确保正确合并传入的参数。更新了相关文件以使用新的合并方式。

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
restore

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Feature: Implement chunk processing and maximum embedding length configuration options to facilitate long-text input support.

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
restore

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
restore

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Feature: Implementation of Chunk Processing for Embedding Requests of Extensive Texts

Signed-off-by: x22x22 <wadeking@qq.com>

DarkLight1337 commented on vllm/inputs/registry.py:
```diff
@@ -242,4 +242,4 @@ def dummy_data_for_profiling(
             multi_modal_data=dec_data.multi_modal_data,
             multi_modal_placeholders=dec_data.multi_modal_placeholders,
-        )
+        )
```
Can you revert these formatting changes?

DarkLight1337 commented on vllm/config.py:
```diff
@@ -3464,6 +3464,25 @@ class PoolerConfig:
+    max_embed_len: Optional[int] = None
+    """
+    Maximum input length allowed for embedding generation. When set, allows 
+    inputs longer than max_model_len to be accepted for embedding models.
```
```suggestion
    inputs longer than max_embed_len to be accepted for embedding models.
```

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +134,490 @@ def _build_response(
+
+        # Get tokenizer config for pooling models (embedding models)
+        if self.model_config.runner_type == "pooling":
+            from vllm.transformers_utils.config import try_get_tokenizer_config
```
Why is this a lazy import?

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +134,490 @@ def _build_response(
+        # Process all chunks for MEAN aggregation
+        chunks_to_process = chunks
+        chunk_indices = list(range(len(chunks)))
+        logger.info("Using chunked processing with MEAN aggregation")
```
Make this a debug log or remove it entirely

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +134,490 @@ def _build_response(
+                length_type = "maximum context length"
+                max_length_value = self.max_model_len
+
+            validation_error_msg = (
```
Use f-string

x22x22 committed:
revert: restore processor.py and registry.py to main branch state

- Restore vllm/transformers_utils/processor.py to main branch
- Restore vllm/inputs/registry.py to main branch
- Ensure all file metadata matches main branch exactly

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactor: Enhance the code structure and error handling logic for embedding generation.

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactor: Enhance the code structure and error handling logic for embedding generation.

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactor: Enhance the code structure and error handling logic for embedding generation.

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Feature: Implementation of an automatic chunking mechanism for long text embedding, accompanied by corresponding unit tests.

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Feature: Implementation of an automatic chunking mechanism for long text embedding, accompanied by corresponding unit tests.

Signed-off-by: x22x22 <wadeking@qq.com>

maxdebayser commented on vllm/entrypoints/openai/serving_engine.py:
```diff
@@ -995,7 +995,9 @@ def _log_inputs(
-            prompt_embeds = inputs.get("prompt_embeds")
+            # Cast to proper type for log_inputs
+            prompt_embeds = cast(Optional[torch.Tensor],
+                                 inputs.get("prompt_embeds"))
```
Is the linter complaining about this? Since you otherwise haven't changed this file, the linter shouldn't complain about typing.

DarkLight1337 replied:
Bump

DarkLight1337 replied:
Any update on this comment?

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +139,505 @@ def _build_response(
+
+        # Process all chunks for MEAN aggregation
+        chunks_to_process = chunks
+        chunk_indices = list(range(len(chunks)))
+
+        for i, (chunk_idx, chunk_tokens) in enumerate(
+                zip(chunk_indices, chunks_to_process)):
```
```suggestion
        for chunk_idx, chunk_tokens in enumerate(chunks_to_process):
```

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +139,505 @@ def _build_response(
+                        "%s, will use chunked processing", token_num,
+                        max_pos_embeddings)
+                else:
+                    raise ValueError(
+                        f"This model's maximum position embeddings length is "
+                        f"{max_pos_embeddings} tokens. However, you requested "
+                        f"{token_num} tokens in the input for embedding "
+                        f"generation. Please reduce the length of the input or "
+                        f"enable chunked processing.")
```
The validation error message can be reused here if you define it as a format string like:
```
            validation_error_msg = (
                "This model's {length_type} is {max_length_value} tokens. "
                "However, you requested {token_num} tokens in the input for "
                "embedding generation. Please reduce the length of the input."
            )

            # Check if input exceeds effective max length
            if token_num > effective_max_len:
                raise ValueError(validation_error_msg.format(length_type=length_type, max_length_value=max_length_value, token_num=token_num))
```

Then below you can do 
```
raise ValueError(validation_error_msg.format(length_type="maximum position embeddings length", max_length_value=max_pos_embeddings, token_num=token_num))
```
https://docs.python.org/3/tutorial/inputoutput.html#the-string-format-method

x22x22 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +139,505 @@ def _build_response(
+        # max_model_len is set in OpenAIServing.__init__
+        # from model_config.max_model_len
+        if self.max_model_len is not None:
+            derived_max_len = min(derived_max_len, self.max_model_len)
```
https://huggingface.co/intfloat/multilingual-e5-large/blob/main/tokenizer_config.json
and
https://huggingface.co/intfloat/multilingual-e5-large/blob/main/config.json

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

noooop commented on examples/online_serving/openai_embedding_long_text/client.py:
```diff
@@ -0,0 +1,368 @@
+   
+   # MEAN pooling (processes all chunks, recommended for complete coverage)
+   vllm serve intfloat/multilingual-e5-large \
+     --task embed \
```
--task embed has been deprecated, it can run normally without this flag

noooop commented on tests/entrypoints/openai/test_embedding_long_text_datasets/long_text_1500_words.txt:
```diff
@@ -0,0 +1,45 @@
+The development of artificial intelligence technology is profoundly transforming our world in unprecedented ways that continue to reshape every aspect of human civilization and society. From sophisticated machine learning algorithms to complex deep neural networks, from advanced natural language processing systems to cutting-edge computer vision models, artificial intelligence technology has demo ...
```
We can randomly generate long texts

x22x22 replied:
Could you please advise on the methodology for generating randomized long-form textual content?

noooop commented on vllm/config.py:
```diff
@@ -5255,4 +5274,4 @@ def update_config(config: DataclassInstanceT,
                 value)
         processed_overrides[field_name] = value
-    return replace(config, **processed_overrides)
+    return replace(config, **processed_overrides)
```
A blank line is needed here.

noooop commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +148,509 @@ def _build_response(
+        """
+        hf_config = self.model_config.hf_config
+
+        # Start with max_position_embeddings from model config
+        derived_max_len = getattr(hf_config, 'max_position_embeddings', 512)
+
+        # Get tokenizer config for pooling models (embedding models)
+        if self.model_config.runner_type == "pooling":
+            tokenizer_config = try_get_tokenizer_config(
+                self.model_config.tokenizer,
+                trust_remote_code=self.model_config.trust_remote_code,
+                revision=self.model_config.tokenizer_revision)
+
+            # Consider model_max_length in tokenizer_config
+            # (same logic as _get_and_verify_max_len)
+            if tokenizer_config:
+                tokenizer_model_max_length = tokenizer_config.get(
+                    'model_max_length', derived_max_len)
+                derived_max_len = min(derived_max_len,
+                                      tokenizer_model_max_length)
+
+        # Consider max_model_len when it's set and smaller than other limits
+        # max_model_len is set in OpenAIServing.__init__
+        # from model_config.max_model_len
+        if self.max_model_len is not None:
+            derived_max_len = min(derived_max_len, self.max_model_len)
+
+        return int(derived_max_len)
```
I think we can directly use model_config.max_model_len, there is no need to copy this code

noooop commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +148,509 @@ def _build_response(
+
+                            # Create a PoolingRequestOutput
+                            # for the aggregated result
+                            from vllm.outputs import PoolingOutput
```
Why is lazy import needed here

noooop reviewed (commented):
The code would be easier to review if it were a bit shorter.

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 commented:
> The code would be easier to review if it were a bit shorter.

@noooop
All recommendations have been duly implemented. Thank you!

hmellor commented on examples/online_serving/openai_embedding_long_text/README.md:
```diff
@@ -0,0 +1,196 @@
+
+## 🚀 Quick Start
+
+### 1. Start the Server
```
Could you remove the explicit numbers from the headings?

hmellor reviewed (commented):
My review is from a docs perspective only.

Cross referencing from examples to other docs is a bit awkward because the depth of the examples is different after they've been converted into pages for `docs.vllm.ai`

x22x22 committed:
Merge branch 'main' into feat/support-long-text-embedding

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 commented:
@DarkLight1337
```
[buildkite/fastcheck/pr/entrypoints-test-api-server](https://buildkite.com/vllm/fastcheck/builds/34788#01989742-66fd-40a2-ae01-2fb75562d623) — Failed (exit status 1)
```
The unit test failure for this API is unrelated to the code I submitted. Could you please advise on the appropriate resolution?

DarkLight1337 commented:
Ignore it, we can force merge if the failures are not related to the PR

x22x22 commented:
> Ignore it, we can force merge if the failures are not related to the PR

@DarkLight1337 
Could you please let me know when my PR can be merged? I've been waiting for almost a month now 😭

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -46,6 +55,15 @@ def _get_embedding(
+        super().__init__(*args, **kwargs)
+
+        # Cache chunked processing support to avoid repeated attribute lookups
+        pooler_config = getattr(self.model_config, 'pooler_config', None)
```
I think `pooler_config` should always be defined on `model_config`?

x22x22 replied:
Actually, [pooler_config](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L858:#L875) is not always defined on [model_config](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L2757:#L2761). Looking at the [ModelConfig](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L2255:#L21731) class definition:

```python
pooler_config: Optional["PoolerConfig"] = field(init=False)
```

The [pooler_config](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L858:875) attribute uses [field(init=False)](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L228:242) and is only initialized when `runner_type == "pooling"` in the [_init_pooler_config()](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L858:875) method. For non-pooling models, it remains `None`.

Using `getattr(self.model_config, 'pooler_config', None)` is the correct defensive programming approach here to safely handle cases where [pooler_config](https://github.com/vllm-project/vllm/blob/ac5b69a056bea0864967241a45e43299dc3b9415/vllm/config/__init__.py#L858:875) might not be initialized, avoiding potential `AttributeError`s when this code runs with different model configurations.

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +147,475 @@ def _build_response(
+            pooler_config = getattr(self.model_config, 'pooler_config', None)
+
+            # Get max_embed_len from pooler config if set
+            max_embed_len = (pooler_config.max_embed_len if pooler_config
```
Can also set this at initialization time

DarkLight1337 commented on vllm/entrypoints/openai/serving_engine.py:
```diff
@@ -1097,7 +1121,12 @@ def _get_decoded_token(logprob: Logprob,
         if logprob.decoded_token is not None:
             return logprob.decoded_token
-        return tokenizer.decode(token_id)
+        # Ensure tokenizer has decode method
```
Is this related to this PR?

DarkLight1337 commented on vllm/entrypoints/openai/serving_engine.py:
```diff
@@ -930,9 +947,14 @@ async def _preprocess_chat(
-            prompt_inputs = TextTokensPrompt(
-                prompt=tokenizer.decode(request_prompt),
-                prompt_token_ids=request_prompt)
+            # Ensure tokenizer has decode method
```
Is this related to this PR?

DarkLight1337 reviewed (commented):
Sorry for the late review, added some comments

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

DarkLight1337 reviewed (commented):
@hmellor feel free to add ready label once my comments have been addressed

x22x22 committed:
Update vllm/entrypoints/openai/serving_embedding.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>
Signed-off-by: Kdump <rootshellexp@gmail.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 commented:
@DarkLight1337 @hmellor
All done, thanks!

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +152,456 @@ def _build_response(
+        token_num = len(input_ids)
+
+        # Note: EmbeddingRequest doesn't have max_tokens
+        if isinstance(request,
+                      (EmbeddingChatRequest, EmbeddingCompletionRequest)):
```
```suggestion
        if isinstance(request, EmbeddingRequest):
```

DarkLight1337 replied:
Using isinstance with a Union type sadly only works in Python 3.10+

maxdebayser replied:
dang :/

x22x22 replied:
```bash
❯ python -m mypy vllm/entrypoints/openai/serving_embedding.py --python-version 3.11
vllm/entrypoints/openai/serving_embedding.py:161: error: Parameterized generics cannot be used with class or instance checks  [misc]
vllm/entrypoints/openai/serving_embedding.py:162: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
vllm/entrypoints/openai/serving_embedding.py:224: error: Parameterized generics cannot be used with class or instance checks  [misc]
vllm/entrypoints/openai/serving_embedding.py:224: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
vllm/entrypoints/openai/serving_embedding.py:381: error: Argument 2 to "_create_single_prompt_generator" of "EmbeddingMixin" has incompatible type "TypedDict({'cache_salt'?: str})"; expected "TokensPrompt | EmbedsPrompt"  [arg-type]
Found 5 errors in 1 file (checked 1 source file)
```

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +152,456 @@ def _build_response(
+            # Determine the effective max length for validation
+            if max_embed_len is not None:
+                # Use max_embed_len for validation instead of max_model_len
+                effective_max_len = max_embed_len
```
`effective_max_len` is not required because it always has the same value as max_lenght_value

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +152,456 @@ def _build_response(
+                "or enable chunked processing.")
+
+            # Check if input exceeds effective max length
+            if token_num > effective_max_len:
```
replace with `max_length_value`

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +152,456 @@ def _build_response(
+            if ctx.engine_prompts is None:
+                return self.create_error_response(
+                    "Engine prompts not available")
+
```
```suggestion
            # Check if we used chunked processing
            use_chunked = self._should_use_chunked_processing(ctx.request)

            if not use_chunked:
                return await super()._collect_batch(ctx=ctx)
```

And then remove the normal processing branch at the end.

x22x22 committed:
Update vllm/entrypoints/openai/serving_embedding.py

Co-authored-by: Maximilien de Bayser <maxdebayser@gmail.com>
Signed-off-by: Kdump <rootshellexp@gmail.com>

x22x22 committed:
Update vllm/entrypoints/openai/serving_embedding.py

Co-authored-by: Maximilien de Bayser <maxdebayser@gmail.com>
Signed-off-by: Kdump <rootshellexp@gmail.com>

x22x22 committed:
Update vllm/entrypoints/openai/serving_embedding.py

Co-authored-by: Maximilien de Bayser <maxdebayser@gmail.com>
Signed-off-by: Kdump <rootshellexp@gmail.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +152,465 @@ def _build_response(
+                ctx.final_res_batch = cast(
+                    list[Union[RequestOutput, PoolingRequestOutput]],
+                    final_res_batch)
+            else:
+                # Normal processing for non-chunked requests
+                num_prompts = len(ctx.engine_prompts)
+                normal_final_res_batch: list[
+                    Optional[PoolingRequestOutput]] = [None] * num_prompts
+
+                async for result_idx, result in ctx.result_generator:
+                    if result_idx < num_prompts:
+                        # Cast to PoolingRequestOutput for embedding results
+                        normal_final_res_batch[result_idx] = cast(
+                            PoolingRequestOutput, result)
+
+                if None in normal_final_res_batch:
+                    return self.create_error_response(
+                        "Failed to generate results for all prompts")
+
+                final_results = [
+                    res for res in normal_final_res_batch if res is not None
+                ]
+                ctx.final_res_batch = cast(
+                    list[Union[RequestOutput, PoolingRequestOutput]],
+                    final_results)
```
This can be removed due to the return above

x22x22 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +152,456 @@ def _build_response(
+
+        # For chunked processing, we always use MEAN aggregation
+        # for cross-chunk aggregation (native pooling is used within each chunk)
+        return self.supports_chunked_processing
```
❯ python -m mypy vllm/entrypoints/openai/serving_embedding.py --python-version 3.11
vllm/entrypoints/openai/serving_embedding.py:161: error: Parameterized generics cannot be used with class or instance checks  [misc]
vllm/entrypoints/openai/serving_embedding.py:162: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
vllm/entrypoints/openai/serving_embedding.py:224: error: Parameterized generics cannot be used with class or instance checks  [misc]
vllm/entrypoints/openai/serving_embedding.py:224: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
vllm/entrypoints/openai/serving_embedding.py:381: error: Argument 2 to "_create_single_prompt_generator" of "EmbeddingMixin" has incompatible type "TypedDict({'cache_salt'?: str})"; expected "TokensPrompt | EmbedsPrompt"  [arg-type]
Found 5 errors in 1 file (checked 1 source file)

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

maxdebayser commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -129,6 +153,472 @@ def _build_response(
+                priority=getattr(ctx.request, "priority", 0),
+            )
+
+            generators.append(generator)
```
Maybe here you can wrap this in a generator that returns a tuple[int, int, Union[RequestOutput, PoolingRequestOutput]], where the first int is the prompt index and the second int is the chunk index. The same can be done in `_create_single_prompt_generator`, but the second int would always be 0. In this way you can eliminate the string splitting logic in `_collect_batch`. It would also remove the need for error handling like in "Invalid chunk request ID format:"

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>

x22x22 commented:
@maxdebayser  @DarkLight1337 @hmellor
All done, thanks!

DarkLight1337 committed:
Reduce diff

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>

DarkLight1337 committed:
Simplify

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>

DarkLight1337 commented:
Let's merge this

DarkLight1337 committed:
Merge branch 'main' into feat/support-long-text-embedding

x22x22 committed:
Refactoring inelegant code

Signed-off-by: x22x22 <wadeking@qq.com>
